### PR TITLE
Zarr chunking fixes

### LINKED ISSUE #2300: zarr and xarray chunking compatibility and `to_zarr` performance
I have a situation where I build large zarr arrays based on chunks which correspond to how I am reading data off a filesystem, for best I/O performance. Then I set these as variables on an xarray dataset which I want to persist to zarr, but with different chunks more optimal for querying.

One problem I ran into is that manually selecting chunks of a dataset prior to `to_zarr` results in https://github.com/pydata/xarray/blob/66be9c5db7d86ea385c3a4cd4295bfce67e3f25b/xarray/backends/zarr.py#L83

It's difficult for me to understand exactly how to select chunks manually at the dataset level which would also make this zarr "final chunk" constraint happy. I would have been satisfied however with letting zarr [choose](https://github.com/zarr-developers/zarr/blob/07fa81a5afe906fcbafc5f81a851c438493ffc13/zarr/util.py#L49) chunks for me, but could not find a way to trigger this through the xarray API short of "unchunking" it first, which would lead to loading entire variables into memory. I came up with the following hack to trigger zarr's automatic chunking despite having differently defined chunks on my xarray dataset:

```python
    # monkey patch to get zarr to ignore dask chunks and use its own heuristics
    def copy_func(f):
        g = types.FunctionType(f.__code__, f.__globals__, name=f.__name__,
                               argdefs=f.__defaults__,
                               closure=f.__closure__)
        g = functools.update_wrapper(g, f)
        g.__kwdefaults__ = f.__kwdefaults__
        return g
    orig_determine_zarr_chunks = copy_func(xr.backends.zarr._determine_zarr_chunks)
    xr.backends.zarr._determine_zarr_chunks = lambda enc_chunks, var_chunks, ndim: orig_determine_zarr_chunks(enc_chunks, None, ndim)
```
The next problem to contend with is that `da.store` between zarr stores with differing chunks between source and destination is astronomically slow. The first thing to attempt would be to rechunk the dask arrays according to the destination zarr chunks, but xarray's [consistent chunks constraint](https://github.com/pydata/xarray/blob/59ad782f29a0f4766bac7802be6650be61f018b8/xarray/core/dataset.py#L1239) blocks this strategy as far as I can tell. Once again I took the dirty hack approach and inject a rechunking on a per-variable basis during the `to_zarr` operation, as follows:

```python
    # monkey patch to make dask arrays writable with different chunks than zarr dest
    # could do without this but would have to contend with 'inconsistent chunks' on dataset
    def sync_using_zarr_copy(self, compute=True):
        if self.sources:
            import dask.array as da
            rechunked_sources = [source.rechunk(target.chunks)
                for source, target in zip(self.sources, self.targets)]
            delayed_store = da.store(rechunked_sources, self.targets,
                                     lock=self.lock, compute=compute,
                                     flush=True)
            self.sources = []
            self.targets = []
            return delayed_store
    xr.backends.common.ArrayWriter.sync = sync_using_zarr_copy
```
I may have missed something in the API that would have made this easier, or another workaround which would be less hacky, but in any case I'm wondering if this scenario could be handled elegantly in xarray.

I'm not sure if there is a plan going forward to make legal xarray chunks 100% compatible with zarr; if so that would go a fair ways in alleviating the first problem. Alternatively, perhaps the xarray API could expose some ability to adjust chunks according to zarr's liking, as well as the option of defaulting entirely to zarr's heuristics for chunking.

As for the performance issue with differing chunks, I'm not sure whether my rechunking patch could be applied without causing side-effects. Or where the right place to solve this would be-- perhaps it could be more naturally addressed within `da.store`.

### LINKED ISSUE #5056: Allow "unsafe" mode for zarr writing
Curently, `Dataset.to_zarr` will only write Zarr datasets in cases in which
- The Dataset arrays are in memory (no dask)
- The arrays are chunked with dask with a one-to-many relationship between dask chunks and zarr chunks

If I try to violate the one-to-many condition, I get an error

```python
import xarray as xr
ds = xr.DataArray([0, 1., 2], name='foo').chunk({'dim_0': 1}).to_dataset()
d = ds.to_zarr('test.zarr', encoding={'foo': {'chunks': (3,)}}, compute=False)
```

```
/srv/conda/envs/notebook/lib/python3.8/site-packages/xarray/backends/zarr.py in _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name)
    148             for dchunk in dchunks[:-1]:
    149                 if dchunk % zchunk:
--> 150                     raise NotImplementedError(
    151                         f"Specified zarr chunks encoding['chunks']={enc_chunks_tuple!r} for "
    152                         f"variable named {name!r} would overlap multiple dask chunks {var_chunks!r}. "

NotImplementedError: Specified zarr chunks encoding['chunks']=(3,) for variable named 'foo' would overlap multiple dask chunks ((1, 1, 1),). This is not implemented in xarray yet. Consider either rechunking using `chunk()` or instead deleting or modifying `encoding['chunks']`.
```

In this case, the error is particularly frustrating because I'm not even writing any data yet. (Also related to #2300, #4046, #4380).

There are at least two scenarios in which we might want to have more flexibility.
1. The case above, when we want to lazily initialize a Zarr array based on a Dataset, without actually computing anything. 
2. The more general case, where we actually write arrays with many-to-many dask-chunk <-> zarr-chunk relationships

For 1, I propose we add a new option like `safe_chunks=True` to `to_zarr`. `safe_chunks=False` would permit just bypassing this chunk.

For 2, we could consider implementing locks. This probably has to be done at the Dask level. But is actually [not super hard](https://github.com/pangeo-forge/pangeo-forge/blob/c42ead11cf2643e815d353637ecb305973b86a53/pangeo_forge/utils.py#L38-L61) to deterministically figure out which chunks need to share a lock.

### PR DESCRIPTION
**Pull request by rabernat**
- [x] Closes #2300, closes #5056
- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

This PR contains two small, related updates to how Zarr chunks are handled.

1. We now delete the `encoding` attribute at the Variable level whenever `chunk` is called. The persistence of `chunk` encoding has been the source of lots of confusion (see #2300,  #4046, #4380, https://github.com/dcs4cop/xcube/issues/347)
2. Added a new option called `safe_chunks` in `to_zarr` which allows for bypassing the requirement of the many-to-one relationship between Zarr chunks and Dask chunks (see #5056).

Both these touch the internal logic for how chunks are handled, so I thought it was easiest to tackle them with a single PR.

### CODE REVIEW

**Comment by rabernat** [comment]
Confused about the test errors. They seem unrelated. In `test_sparse.py:test_variable_method`

```
E   TypeError: no implementation found for 'numpy.allclose' on types that implement __array_function__: [<class 'numpy.ndarray'>, <class 'sparse._coo.core.COO'>]
```

**Comment by andersy005** [comment]
> Confused about the test error. It seems unrelated. In `test_sparse.py:test_variable_method`
> 
> ```
> E   TypeError: no implementation found for 'numpy.allclose' on types that implement __array_function__: [<class 'numpy.ndarray'>, <class 'sparse._coo.core.COO'>]
> ```


Related to #5059, and it appears that @keewis came up with a fix for it in https://github.com/pydata/xarray/pull/5059#issuecomment-803620128

**Comment by rabernat** [comment]
Thanks Anderson. Fixed by rebasing. Now RTD build is failing, but there is no obvious error in the logs...

File: xarray/backends/zarr.py [hunk 1]
```diff
@@ -588,7 +613,7 @@ def open_zarr(
         values according to the formula `original_values * scale_factor +
         add_offset`, where `_FillValue`, `scale_factor` and `add_offset` are
         taken from variable attributes (if they exist).  If the `_FillValue` or
-        `missing_value` attribute contains multiple values a warning will be
+            `missing_value` attribute contains multiple values a warning will be
```
Comment (keewis) [inline xarray/backends/zarr.py:616]: the errors are:
```
/home/docs/checkouts/readthedocs.org/user_builds/xray/checkouts/5065/xarray/backends/zarr.py:docstring of xarray.backends.zarr.open_zarr:28: WARNING: Unexpected indentation.
/home/docs/checkouts/readthedocs.org/user_builds/xray/checkouts/5065/xarray/backends/zarr.py:docstring of xarray.backends.zarr.open_zarr:29: WARNING: Block quote ends without a blank line; unexpected unindent.
```

my guess is that both would be fixed by undoing this:
```suggestion
        `missing_value` attribute contains multiple values a warning will be
```

File: xarray/core/dataset.py [hunk 2]
```diff
@@ -1779,6 +1789,11 @@ def to_zarr(
               in with ``region``, use a separate call to ``to_zarr()`` with
               ``compute=False``. See "Appending to existing Zarr stores" in
               the reference documentation for full details.
+        safe_chunks : bool, optional
+            If True, only allow writes to when there is a many-to-one relationship
+            between Zarr chunks and Dask chunks. Set False to override this
+            restriction; however, data may become corrupted if Zarr arrays
+            are written in paralle.
```
Comment (dcherian) [inline xarray/core/dataset.py:1796]: Let's add a line about a possible use-case for this option 

From the issue,
> to lazily initialize a Zarr array based on a Dataset, without actually computing anything.

Comment (dcherian) [inline xarray/core/dataset.py:1796]:
```suggestion
            are written in parallel.
```

File: xarray/backends/zarr.py [hunk 3]
```diff
@@ -163,24 +164,36 @@ def _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name):
...
+                    base_error = (
                         f"Specified zarr chunks encoding['chunks']={enc_chunks_tuple!r} for "
                         f"variable named {name!r} would overlap multiple dask chunks {var_chunks!r}. "
-                        "This is not implemented in xarray yet. "
-                        "Consider either rechunking using `chunk()` or instead deleting "
-                        "or modifying `encoding['chunks']`."
+                        f"Writing this array in parallel with dask could lead to corrupted data. "
                     )
+                    if safe_chunks:
+                        raise ValueError(
```
Comment (dcherian) [inline xarray/backends/zarr.py:173]: really minor comment: Shouldn't this still be `NotImplementedError` since we could technically support this by implementing locks?

File: doc/whats-new.rst [hunk 4]
```diff
@@ -23,6 +23,9 @@ v0.17.1 (unreleased)
 New Features
 ~~~~~~~~~~~~
 
+- Add `safe_chunks` option to :py:meth:`Dataset.to_zarr` which allows overriding
```
Comment (dcherian) [inline doc/whats-new.rst:26]:
```suggestion
- Add ``safe_chunks`` option to :py:meth:`Dataset.to_zarr` which allows overriding
```

File: doc/whats-new.rst [hunk 5]
```diff
@@ -74,6 +77,8 @@ Deprecations
 
 Bug fixes
 ~~~~~~~~~
+- `chunk` methods now delete variable `chunk` encoding (:issue:`2300`)
```
Comment (dcherian) [inline doc/whats-new.rst:80]:
```suggestion
- :py:meth:`Dataset.chunk` and :py:meth:`DataArray.chunk` methods now delete variable ``chunks`` encoding (:issue:`2300`)
```

File: xarray/core/variable.py [hunk 6]
```diff
@@ -1091,6 +1092,10 @@ def chunk(self, chunks={}, name=None, lock=False):
 
             data = da.from_array(data, chunks, name=name, lock=lock, **kwargs)
 
+        # rechunking erases encoding
+        if self._encoding and "chunks" in self._encoding:
+            del self._encoding["chunks"]
```
Comment (dcherian) [inline xarray/core/variable.py:1097]: we should mention this in the docstring for `DataArray.chunk` `Dataset.chunk` and `Variable.chunk`

**Review by dcherian — COMMENTED** [review]
Thanks @rabernat I only have some docstring suggestions

File: xarray/backends/zarr.py (hunk 3 above)
Comment (rabernat) [reply xarray/backends/zarr.py:173]: It used to be, but I changed it! Do we ever plan to implement locks?

**Comment by shoyer** [comment]
I'm a little conflicted about dealing with `encoding['chunks']` specifically in `chunk()`:

- On one hand, it feels inconsistent for this only this single method in xarray to modify part of `encoding`. Nothing else in xarray (after CF decoding) does this. Effectively `encoding['chunks']` is now becoming a part of xarray's data model.
- On the other hand, this would absolutely fix a recurrent pain-point for users, and in that sense it's worth doing.

Maybe this isn't such a big deal in this particular case, especially if we don't think we would need to add such encoding specific logic to any other methods. But are we really sure about that -- what about cases like indexing?

I guess the other alternative to make `chunk()` and various other methods that would change chunking drop `encoding` entirely. I don't know if this would really be a better comprehensive solution (I know dropping `attrs` is much hated), but at least it's an easier mental model.

**Comment by rabernat** [comment]
I see you point. I guess I don't fully understand where else in the code path encoding gets dropped. Consider this example

```python
import xarray as xr
ds = xr.Dataset({'foo': ('time', [1, 1], {'dtype': 'int16'})})
ds = xr.decode_cf(ds).compute()
assert "dtype" in ds.foo.encoding
assert "dtype" not in (0.5 * ds.foo).encoding
```

Xarray knows to drop the `dtype` encoding after an arithmetic operation. How does that work? To me `.chunk` feel like a similar case: an operation that invalidates any existing encoding.

**Comment by shoyer** [comment]
> Xarray knows to drop the `dtype` encoding after an arithmetic operation. How does that work? To me `.chunk` feel like a similar case: an operation that invalidates any existing encoding.

To be honest, the existing convention is quite adhoc, just based on what seemed most appropriate at the time.

https://github.com/pydata/xarray/issues/1614 is most comprehensive description of the current state of things.

We were considering saying that `attrs` and `encoding` should always use the same rules, but perhaps we should be more aggressive about dropping `encoding`.

**Comment by dcherian** [comment]
> Xarray knows to drop the dtype encoding after an arithmetic operation. How does that work? 

There's a subtle difference. It drops all of `.encoding` not `dtype` specifically.

@shoyer's point about indexing changing chunking is a good one too. Perhaps a kwarg in `to_zarr` like `ignore_encoding_chunks`?

**Comment by rabernat** [comment]
> Perhaps a kwarg in `to_zarr` like `ignore_encoding_chunks`?

I would argue that this is unnecessary. If you want to explicitly drop encoding, just `del da.encoding['chunks']` before writing. But most users don't figure out that they should do this, because the default behavior is counterintuitive.

The problem here is with the default behavior of propagating chunk encoding through computations when it no longer makes sense. My example with the `dtype` encoding illustrates that we already drop encoding on certain operations, so it's not unprecedented. It's more of an implementation question: where and how to do the dropping.

FWIW, I would also favor dropping `encoding['chunks']` after indexing, coarsening, interpolating, etc. Basically anything that changes the array shape or chunk structure.

**Comment by shoyer** [comment]
> FWIW, I would also favor dropping `encoding['chunks']` after indexing, coarsening, interpolating, etc. Basically anything that changes the array shape or chunk structure.

We already drop all of `encoding` after indexing. My guess is that we do the same for coarsening and interpolations as well (though I haven't checked).

**Comment by aurghs** [comment]
Perhaps we could remove also ``overwrite_encoded_chunks``, it shouldn't be any more necessary.

> The problem here is with the default behavior of propagating chunk encoding through computations when it no longer makes sense. My example with the `dtype` encoding illustrates that we already drop encoding on certain operations, so it's not unprecedented. It's more of an implementation question: where and how to do the dropping.

Maybe, in this case, the right moment to do the dropping is at reading the data. If the user doesn't re-chunk the data, the encoded chunks information is duplicated in dask-chunks.  
We could add a key in ``to_zarr``interface: ``ds.to_zarr(..., chunks=...)`` for specifying a different chunk to be used for writing the data.

**Comment by rabernat** [comment]
I appreciate the discussion on this PR. Does anyone have a concrete suggestion of what to do?

If we are not in agreement about the encoding stuff, perhaps I should remove that and just move forward with the `safe_chunks` part of this PR?

**Comment by rabernat** [comment]
In today's dev call, we proposed to handle encoding in `chunk` the same way we handle it in indexing: by deleting all encoding. 

The problem is, I can't figure out where this happens. Can someone point me to the place in the code where indexing operations delete encoding?

A related question: I discovered this encoding option `preferred_chunk`, which is treated specially:
https://github.com/pydata/xarray/blob/57a4479fcd3ebc579cf00e0d6bf85007eda44b56/xarray/core/dataset.py#L396

Should the Zarr backend be setting this?

**Comment by aurghs** [comment]
> Should the Zarr backend be setting this?

Yes.  We decide to separate the `chunks` and the `preferred_chunks`:
- The `preferred_chunks` is used by the backend to define the deafult chunks to be used by xarray. 
- The `chunks` are the on-disk chunks.

They are not necessarily the same. 
Maybe we can drop the `preferred_chunks` can be dropped after they are used.

**Comment by aurghs** [comment]
`rechunk` is used always when you open a data with dask, even if you are using the default chunking. So in this way, you will drop the encoding always when dask is used (≈ always).

**Comment by dcherian** [comment]
> The problem is, I can't figure out where this happens. 

Replace `self._encoding` with `None` here? https://github.com/pydata/xarray/blob/ddc352faa6de91f266a1749773d08ae8d6f09683/xarray/core/variable.py#L1084

**Comment by rabernat** [comment]
> Replace `self._encoding` with `None` here?

Thanks! Yeah that's what I had in mind. But I was wondering if there was an example of doing that it else I could copy.

In any case, I'll give it a try now.

**Comment by rabernat** [comment]
A just pushed a new commit which deletes all encoding after chunk. But as you will see when the CI finishes, this leads to a lot of test failures. For example:

```
=============================================================================== FAILURES ================================================================================
____________________________________________________ TestNetCDF4ViaDaskData.test_roundtrip_string_encoded_characters ____________________________________________________

self = <xarray.tests.test_backends.TestNetCDF4ViaDaskData object at 0x18cba4c40>

    def test_roundtrip_string_encoded_characters(self):
        expected = Dataset({"x": ("t", ["ab", "cdef"])})
        expected["x"].encoding["dtype"] = "S1"
        with self.roundtrip(expected) as actual:
            assert_identical(expected, actual)
>           assert actual["x"].encoding["_Encoding"] == "utf-8"
E           KeyError: '_Encoding'

/Users/rpa/Code/xarray/xarray/tests/test_backends.py:485: KeyError
```

Why is `chunk` getting called here? Does it actually get called every time we load a dataset with chunks? If so, we will need a more sophisticated solution.

**Comment by aurghs** [comment]
>  Does it actually get called every time we load a dataset with chunks? 
Yes

**Comment by rabernat** [comment]
So any ideas how to proceed? 🧐

**Comment by shoyer** [comment]
Hmm. I would also be happy with explicitly deleting `chunks` from encoding for now. It's not adding a lot of technical debt.

In the long term, the whole handling of encoding should be revisited, e.g., see https://github.com/pydata/xarray/issues/5082

File: xarray/core/variable.py [hunk 7]
```diff
@@ -1091,7 +1092,8 @@ def chunk(self, chunks={}, name=None, lock=False):
 
             data = da.from_array(data, chunks, name=name, lock=lock, **kwargs)
 
-        return type(self)(self.dims, data, self._attrs, self._encoding, fastpath=True)
+        new_encoding = None  # rechunking removes all encoding
+        return type(self)(self.dims, data, self._attrs, new_encoding, fastpath=True)
```
Comment (shoyer) [inline xarray/core/variable.py:1096]: a simpler way to achieve the same thing is just to omit the argument:
```suggestion
        return type(self)(self.dims, data, self._attrs, fastpath=True)
```

**Comment by shoyer** [comment]
> Why is `chunk` getting called here? Does it actually get called every time we load a dataset with chunks? If so, we will need a more sophisticated solution.

This happens specifically on this line:
https://github.com/pydata/xarray/blob/ddc352faa6de91f266a1749773d08ae8d6f09683/xarray/core/dataset.py#L438

So perhaps it would make sense to copy `encoding` specifically in this case, e.g.,
```python
        new_var = var.chunk(chunks, name=name2, lock=lock)
        new_var.encoding = var.encoding
```

**Comment by aurghs** [comment]
> ```python
>         new_var = var.chunk(chunks, name=name2, lock=lock)
>         new_var.encoding = var.encoding
> ```
Here you are modifying `_maybe_chunk`, but `_maybe_chunk` is also used in `Dataset.chunks`. 
Probably would be better to change `api.py`, here:
https://github.com/pydata/xarray/blob/ddc352faa6de91f266a1749773d08ae8d6f09683/xarray/backends/api.py#L296-L307

But maybe also in this case we want to drop `encoding["chunks"]` if they are not compatible with dask ones. We could use
https://github.com/pydata/xarray/blob/ddc352faa6de91f266a1749773d08ae8d6f09683/xarray/core/dataset.py#L363
but some small changes are needed.

**Comment by aurghs** [comment]
> Hmm. I would also be happy with explicitly deleting `chunks` from encoding for now. It's not adding a lot of technical debt.
 
I see two reasons for keeping it:
- We should be able to read and write the data with the same structure on file.
- We the user may be interested in this information.

But it seems to me that having two different definitions of chunks (dask one and encoded one), is not very intuitive and it's not easy to define a clear default in writing.

**Comment by rabernat** [comment]
> But it seems to me that having two different definitions of chunks (dask one and encoded one), is not very intuitive and it's not easy to define a clear default in writing.

My use for `encoding.chunks` is to tell Zarr what chunks to use on disk.

**Comment by rabernat** [comment]
We seem to be unable to resolve the complexities around chunk encoding. I propose to remove this from the PR and reduce the scope to just the `safe_chunks` features. @aurghs should probably be the one to tackle the chunk encoding problem; unfortunately it exceeds my understanding, and I don't have time to dig deeper at the moment. In the meantime `safe_chunks` is important for pangeo-forge forward progress.

Please give a 👍  or 👎  to this idea if you have an opinion.

**Comment by rabernat** [comment]
I have removed the controversial `encoding['chunks']` stuff from the PR. Now it only contains the `safe_chunks` option in `to_zarr`.

If there are no further comments on this, I think this is good to go.

File: xarray/backends/zarr.py (hunk 1 above)
Comment (keewis) [reply xarray/backends/zarr.py:616]: not sure how, but this seems to have been reintroduced and causes the docs build to fail.

**Comment by rabernat** [comment]
Any further feedback on this now reduced-scope PR? Merging this would be helpful for moving forward Pangeo forge.

File: xarray/backends/zarr.py [hunk 8]
```diff
@@ -163,24 +164,36 @@ def _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name):
...
             for dchunk in dchunks[:-1]:
                 if dchunk % zchunk:
-                    raise NotImplementedError(
+                    base_error = (
                         f"Specified zarr chunks encoding['chunks']={enc_chunks_tuple!r} for "
                         f"variable named {name!r} would overlap multiple dask chunks {var_chunks!r}. "
-                        "This is not implemented in xarray yet. "
-                        "Consider either rechunking using `chunk()` or instead deleting "
-                        "or modifying `encoding['chunks']`."
+                        f"Writing this array in parallel with dask could lead to corrupted data. "
```
Comment (keewis) [inline xarray/backends/zarr.py:170]:
```suggestion
                        f"Writing this array in parallel with dask could lead to corrupted data."
```

File: xarray/backends/zarr.py [hunk 9]
```diff
@@ -163,24 +164,36 @@ def _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name):
...
+                        warnings.warn(base_error)
             if dchunks[-1] > zchunk:
-                raise ValueError(
+                base_error = (
                     "Final chunk of Zarr array must be the same size or "
                     "smaller than the first. "
                     f"Specified Zarr chunk encoding['chunks']={enc_chunks_tuple}, "
                     f"for variable named {name!r} "
                     f"but {dchunks} in the variable's Dask chunks {var_chunks} is "
                     "incompatible with this encoding. "
```
Comment (keewis) [inline xarray/backends/zarr.py:187]:
```suggestion
                    "incompatible with this encoding."
```

File: xarray/core/dataset.py [hunk 10]
```diff
@@ -1722,12 +1722,22 @@ def to_zarr(
         consolidated: bool = False,
         append_dim: Hashable = None,
         region: Mapping[str, slice] = None,
+        safe_chunks: bool = True,
     ) -> "ZarrStore":
         """Write dataset contents to a zarr group.
 
-        .. note:: Experimental
-                  The Zarr backend is new and experimental. Please report any
-                  unexpected behavior via github issues.
```
Comment (keewis) [inline xarray/core/dataset.py:1730]: is this something we should do in a separate PR, or should we remove all similar warnings here, too? Also, should we add a entry to `whats-new.rst` for this?

File: xarray/backends/zarr.py [hunk 11]
```diff
@@ -163,24 +164,36 @@ def _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name):
...
-                    "Consider either rechunking using `chunk()` or instead deleting "
-                    "or modifying `encoding['chunks']`."
                 )
+                if safe_chunks:
+                    raise NotImplementedError(
+                        base_error
+                        + "Consider either rechunking using `chunk()`, deleting "
```
Comment (keewis) [inline xarray/backends/zarr.py:192]:
```suggestion
                        + " Consider either rechunking using `chunk()`, deleting "
```

File: xarray/backends/zarr.py [hunk 12]
```diff
@@ -163,24 +164,36 @@ def _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name):
...
                         f"variable named {name!r} would overlap multiple dask chunks {var_chunks!r}. "
-                        "This is not implemented in xarray yet. "
-                        "Consider either rechunking using `chunk()` or instead deleting "
-                        "or modifying `encoding['chunks']`."
+                        f"Writing this array in parallel with dask could lead to corrupted data. "
                     )
+                    if safe_chunks:
+                        raise NotImplementedError(
+                            base_error
+                            + "Consider either rechunking using `chunk()`, deleting "
```
Comment (keewis) [inline xarray/backends/zarr.py:175]:
```suggestion
                            + " Consider either rechunking using `chunk()`, deleting "
```

**Review by keewis — COMMENTED** [review]
A few documentation issues, but otherwise looks good to me. I don't know a lot about chunking and zarr, though.

File: xarray/core/dataset.py (hunk 10 above)
Comment (rabernat) [reply xarray/core/dataset.py:1730]: I have updated whats-new.

I agree that we should remove the notes about Zarr be experimental.

File: xarray/backends/zarr.py [hunk 13]
```diff
@@ -163,24 +164,36 @@ def _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name):
...
+                    "incompatible with this encoding."
                 )
+                if safe_chunks:
+                    raise NotImplementedError(
+                        base_error
+                        + " Consider either rechunking using `chunk()`, deleting "
+                        "or modifying `encoding['chunks']`, or specify `safe_chunks=False`."
+                    )
+                else:
+                    warnings.warn(base_error)
```
Comment (dcherian) [inline xarray/backends/zarr.py:196]: Seems weird to warn when the user has opted in to unsafe behaviour

```suggestion
```

File: xarray/backends/zarr.py [hunk 14]
```diff
@@ -163,24 +164,36 @@ def _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name):
...
+                        f"Writing this array in parallel with dask could lead to corrupted data."
                     )
+                    if safe_chunks:
+                        raise NotImplementedError(
+                            base_error
+                            + " Consider either rechunking using `chunk()`, deleting "
+                            "or modifying `encoding['chunks']`, or specify `safe_chunks=False`."
+                        )
+                    else:
+                        warnings.warn(base_error)
```
Comment (dcherian) [inline xarray/backends/zarr.py:179]:
```suggestion
```
(as below)

**Review by dcherian — APPROVED** [review]
Thanks @rabernat

**Comment by rabernat** [comment]
The pre-commit workflow is raising a blackdoc error I am not seeing in my local env

```diff
diff --git a/doc/internals/duck-arrays-integration.rst b/doc/internals/duck-arrays-integration.rst
index eb5c4d8..2bc3c1f 100644
--- a/doc/internals/duck-arrays-integration.rst
+++ b/doc/internals/duck-arrays-integration.rst
@@ -25,7 +25,7 @@ argument:
         ...
 
         def _repr_inline_(self, max_width):
-            """ format to a single line with at most max_width characters """
+            """format to a single line with at most max_width characters"""
             ...
 ```

**Comment by keewis** [comment]
the reason is that `black` released a new version yesterday, and since we don't pin `black` for the `blackdoc` entry we get the new version. If you run `pre-commit clean` before `pre-commit run --all-files` you should see this change locally, too. To avoid situations like these we could start pinning `black` in the `blackdoc` entry (and run a script to synchronize this with the `black` entry on autoupdate).

**Comment by rabernat** [comment]
I think this PR has received a very thorough review. I would be pleased if someone from @pydata/xarray would merge it soon.

**Comment by dcherian** [comment]
Thanks @rabernat
